### ISSUE TITLE
Do we want to create nested objects in the JSON output where a row describes multiple resources?

### ISSUE THREAD

**Issue by 6a6d74** [w3c/csvw#299]
Nesting of objects in the JSON output relies on recursion through the objects defined for each subject within a row in order to "plug" related objects together in a hierarchy. 

This adds a little complexity - but is worth doing (I think) to create graph-like output familiar to most JSON users. Without this recursion, the JSON output would just be a flat set of objects.

For example, the music event listing example that creates 3 objects per row (event, location and offer) creates the following output _WITH_ nesting:

```
[{
  "@id": "http://example.org/events-listing.csv#event-1",
  "rdf:type": "schema:MusicEvent",
  "schema:name": "B.B. King",
  "schema:startDate": "2014-04-12T19:30:00",
  "schema:location": {
    "@id": "http://example.org/events-listing.csv#place-1",
    "rdf:type": "schema:Place",
    "schema:name": "Lupo’s Heartbreak Hotel",
    "schema:address": "79 Washington St., Providence, RI"
  },
  "schema:offer": {
    "@id": "http://example.org/events-listing.csv#offer-1",
    "rdf:type": "schema:Offer",
    "schema:offer": "https://www.etix.com/ticket/1771656"
  }
}, {
  "@id": "http://example.org/events-listing.csv#event-2",
  "rdf:type": "schema:MusicEvent",
  "schema:name": "B.B. King",
  "schema:startDate": "2014-04-13T20:00:00",
  "schema:location": {
    "@id": "http://example.org/events-listing.csv#place-2",
    "rdf:type": "schema:Place",
    "schema:name": "Lynn Auditorium",
    "schema:address": "Lynn, MA, 01901"
  },
  "schema:offer": {
    "@id": "http://example.org/events-listing.csv#offer-2",
    "rdf:type": "schema:Offer",
    "schema:offer": "http://frontgatetickets.com/venue.php?id=11766"
  }
}]
```

Yet without nesting we get a flat object list - which is not very friendly as consuming applications need to assemble the graph themselves based on the `@id` value; e.g.:

```
[{
  "@id": "http://example.org/events-listing.csv#event-1",
  "rdf:type": "schema:MusicEvent",
  "schema:name": "B.B. King",
  "schema:startDate": "2014-04-12T19:30:00",
  "schema:location": "http://example.org/events-listing.csv#place-1",
  "schema:offer": "http://example.org/events-listing.csv#offer-1"
}, {
  "@id": "http://example.org/events-listing.csv#place-1",
  "rdf:type": "schema:Place",
  "schema:name": "Lupo’s Heartbreak Hotel",
  "schema:address": "79 Washington St., Providence, RI"
}, {
  "@id": "http://example.org/events-listing.csv#offer-1",
  "rdf:type": "schema:Offer",
  "schema:offer": "https://www.etix.com/ticket/1771656"
}, {
  "@id": "http://example.org/events-listing.csv#event-2",
  "rdf:type": "schema:MusicEvent",
  "schema:name": "B.B. King",
  "schema:startDate": "2014-04-13T20:00:00",
  "schema:location": "http://example.org/events-listing.csv#place-2",
  "schema:offer": "http://example.org/events-listing.csv#offer-2"
}, {
  "@id": "http://example.org/events-listing.csv#place-2",
  "rdf:type": "schema:Place",
  "schema:name": "Lynn Auditorium",
  "schema:address": "Lynn, MA, 01901"
}, {
  "@id": "http://example.org/events-listing.csv#offer-2",
  "rdf:type": "schema:Offer",
  "schema:offer": "http://frontgatetickets.com/venue.php?id=11766"
}]
```

Personally, I prefer the nested option. This is only possible where objects are referenced once and only once via the `valueUrl` in each row - in all other situations we have to revert back to a flat list. The algorithm in csv2json proposes a (recursive) method to do this:

> If more than one subject is associated with the current row, review each of the objects associated with those subjects to determine whether they can be nested.
> 
> For each subject _i_ associated with the current row:
> 
> ... Recurse through the name-value pairs in the objects associated with the current row to determine if subject _i_ is referenced - excluding the object describing subject _i_ itself: object _Si_.
> 
> ... If subject _i_ is referenced once _and only once_, then object _Si_ SHALL be inserted into the value of the referring name-value pair in place of the reference in order to create a nested object.
> 
> ... Else, object _Si_, potentially including nested objects, SHALL be inserted into [the containing array].

**Comment by iherman** [comment]
I am not sure I understand the "- excluding the object describing subject i itself" remark. We are talking about the references to the objects, so I am not sure that exception occurs.

Otherwise it works, I believe...

**Comment by 6a6d74** [comment]
@iherman - I'm trying to say "don't check the object that represents the _current_ subject" ... we need to avoid self-referencing loops :-)

**Comment by 6a6d74** [comment]
@iherman - I'm trying to say "don't check the object that represents the _current_ subject" ... we need to avoid self-referencing loops :-)

**Comment by iherman** [comment]
Ah, you mean except the object where "@id" is the current subject. My RDF-head did not even think about this one, but you're right

Ivan

> On 05 Mar 2015, at 10:33 , Jeremy Tandy notifications@github.com wrote:
> 
> @iherman - I'm trying to say "don't check the object that represents the current subject" ... we need to avoid self-referencing loops :-)
> 
> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by 6a6d74** [comment]
@iherman - that's what I mean. Feel free to suggest alternative text via a PR.

**Comment by iherman** [comment]
Having gone through the text, it is convoluted, but I do not really know better:-)

**Comment by 6a6d74** [comment]
Thanks ...

**Comment by JeniT** [comment]
I'm not sure it completely works: what if you have an aboutUrl A that has a property that references a valueUrl B, but B is also used as an aboutUrl for a column whose valueUrl is A? I think the following rough step of steps works:
1. make a de-duplicated list of URLs that appear as `aboutUrl`s in the row, in order that they appear in the columns, including `null` if some columns don't have an `aboutUrl`; this is the _URL list_
2. create an object for each URL in the _URL list_; give these objects `@id` properties as relevant (the object for the `null` URL won't have an `@id` property)
3. work out which `valueUrl`s are used within the row and associate each with a count of the number of columns for which that is the `valueUrl`; this is the _value URL list_
4. select a URL from _URL list_ and remove it from that list; this should be, in order of preference:
   1. the `null` URL
   2. the first URL in _URL list_ that doesn't appear in the _value URL list_
   3. the first URL in _URL list_ that has an associated count in the _value URL list_ of more than one
   4. the first URL in _URL list_
5. work through each of the columns whose `aboutUrl` is that URL, adding properties to its associated object
6. if a column has a `valueUrl` then:
   1. if the count associated with the `valueUrl` in the _value URL list_ is more than one, or if the `valueUrl` isn't in _URL list_, just use the `valueUrl` as the value of the property
   2. otherwise, use the object for that `valueUrl` as the value of the property, remove that URL from _URL list_, and repeat steps 5 & 6 for that URL
7. if the _URL list_ still has items in it, go to step 4

**Comment by iherman** [comment]
@JeniT @6a6d74 : I _think_ this works and avoid a pitfall in the current spec. 

However, I am a bit afraid of the complexity. What about 
- Saying that, by default, the JSON conversion produces a simple, 'flat' output
- That processors may have the option of introducing a "smart" option, adding the steps above in an appendix and refer to that as a possible algorithm to follow

Even if the group decision is to include those steps as normative steps, we should be careful how to include them editorially; just adding those points into the current structure will make the spec hardly readable (it is already deeply nested enough). Ie, the algorithm should probably be put into its separate section and referred to from section 3.2

**Comment by JeniT** [comment]
The simple 'flat' option is really odd from a JSON perspective: I expect people who are processing JSON created out of this process to ignore the `@id`s much of the time. It would make the output from examples like the one at http://w3c.github.io/csvw/csv2json/#example-events-listing much less usable for developers. I prefer complexity for implementers than for the users of the results of the conversion.

Yes, the algorithm could do with editorial work to make it readable. I think using a "call out" style (having different named "functions", as you would if you were programming) works better than a deeply nested structure.

**Comment by iherman** [comment]
> The simple 'flat' option is really odd from a JSON perspective: I expect people who are processing JSON created out of this process to ignore the @ids much of the time. It would make the output from examples like the one at http://w3c.github.io/csvw/csv2json/#example-events-listing much less usable for developers. I prefer complexity for implementers than for the users of the results of the conversion.
> 
> Ok.
> Yes, the algorithm could do with editorial work to make it readable. I think using a "call out" style (having different named "functions", as you would if you were programming) works better than a deeply nested structure.
> 
> Yes. We will have to look at that

Ivan

**Comment by 6a6d74** [comment]
Re-reading through the posts for this ISSUE, I think that there is still work to do for the csv2json conversion algorithm? 

Do we need to bottom this out _before_ releasing the next WD? 

If yes, then I'd appreciate some guidance with how to amend the csv2json conversion algorithm ... @JeniT suggested amendments and use of a "call out" style (having different named "functions") - but I am not sure what that actually means ... 

Thanks.

**Comment by iherman** [comment]
> On 09 Mar 2015, at 09:43 , Jeremy Tandy notifications@github.com wrote:
> 
> Re-reading through the posts for this ISSUE, I think that there is still work to do for the csv2json conversion algorithm?

Yes:-(

> Do we need to bottom this out before releasing the next WD?

I think so. The current algorithm in the document has a flaw, as @JeniT pointed out, insofar as it can lead to infinite loops.

> If yes, then I'd appreciate some guidance with how to amend the csv2json conversion algorithm ... @JeniT suggested amendments and use of a "call out" style (having different named "functions") - but I am not sure what that actually means ...

I think the idea is to have a separate section in the document entitled something like "Json nesting algorithm", or something similar, that would contain her algorithm and would return a bunch of JSON objects, and the main flow of the text would refer to that section for the details as for how to generate the right json objects.

I am, at this moment, at Boston airport, en route to Toronto. I can definitely not do anything with this today. I _may_ have some time tomorrow, but I am not sure; I have a conference on Wednesday and I am then on my way back home starting Thursday. Ie, I am not sure I can look into this this week. Whoever is first:-)

Ivan

> Thanks.
> 
> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by gkellogg** [comment]
I plan to implement it shortly, which should reveal some issues. No real feedback before then.

**Comment by 6a6d74** [comment]
@iherman - thanks for the pointers.
@gkellogg - I won't do anything before receiving your implementation-derived feedback; thanks!

**Comment by JeniT** [comment]
@6a6d74 by "call out style" I mean something like what I've written for the model document here: http://w3c.github.io/csvw/syntax/#parsing where there are sub-algorithms (like function calls) named things like **read a row** or **trim a cell value**.

**Comment by gkellogg** [comment]
The process @JeniT outlines is similar to that used by my various RDF serializers which try to embed objects together. Indeed, there is the potential for looping endlessly, but I think it can be solved by doing multiple passes and stopping when the set of remaining objects remains constant.

Another possible issue with the algorithm is if multiple cells have the same propertyUrl, as this would be another case when an object property would have multiple values without the use of separator.

I think the solution is similar to that used for JSON-LD framing, where you start by building up a flattened structure by processing each cell, creating an object with `@id` from the aboutUrl as necessary, and adding properties and appending values as encountered.

Subsequently, use an algorithm similar to @JeniT's to identify top-level objects who's `@id` is also a value within some other object and moving replacing the reference node with the top-level object. This requires recursing down through objects to find node references (i.e. `{"@id": "ref"}`) that match a top-level object. The process can repeat until the number of remaining top-level objects remains the same. if it ends in a single object (as it typically will), the array containing that object can be replaced with just the object.

This certainly should be described in a sub-algorithm (call out). I'm happy to work on expressing such an algorithm during the course of my implementation.

**Comment by iherman** [comment]
> On 11 Mar 2015, at 02:10 , Gregg Kellogg notifications@github.com wrote:
> 
> The process @JeniT outlines is similar to that used by my various RDF serializers which try to embed objects together. Indeed, there is the potential for looping endlessly, but I think it can be solved by doing multiple passes and stopping when the set of remaining objects remains constant.
> 
> Another possible issue with the algorithm is if multiple cells have the same propertyUrl, as this would be another case when an object property would have multiple values without the use of separator.

Let us forget about the issue of nesting objects: this is an issue even if all columns use the same aboutUrl, and this is not handled by the current text!

I will add a separate issue for this

> I think the solution is similar to that used for JSON-LD framing, where you start by building up a flattened structure by processing each cell, creating an object with @id from the aboutUrl as necessary, and adding properties and appending values as encountered.
> 
> Subsequently, use an algorithm similar to @JeniT's to identify top-level objects who's @id is also a value within some other object and moving replacing the reference node with the top-level object. This requires recursing down through objects to find node references (i.e. {"@id": "ref"}) that match a top-level object. The process can repeat until the number of remaining top-level objects remains the same. if it ends in a single object (as it typically will), the array containing that object can be replaced with just the object.
> 
> This certainly should be described in a sub-algorithm (call out). I'm happy to work on expressing such an algorithm during the course of my implementation.
> 
> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by iherman** [comment]
> On 11 Mar 2015, at 02:10 , Gregg Kellogg notifications@github.com wrote:
> 
> The process @JeniT outlines is similar to that used by my various RDF serializers which try to embed objects together. Indeed, there is the potential for looping endlessly, but I think it can be solved by doing multiple passes and stopping when the set of remaining objects remains constant.
> 
> Another possible issue with the algorithm is if multiple cells have the same propertyUrl, as this would be another case when an object property would have multiple values without the use of separator.
> 
> I think the solution is similar to that used for JSON-LD framing, where you start by building up a flattened structure by processing each cell, creating an object with @id from the aboutUrl as necessary, and adding properties and appending values as encountered.
> 
> Subsequently, use an algorithm similar to @JeniT's to identify top-level objects who's @id is also a value within some other object and moving replacing the reference node with the top-level object. This requires recursing down through objects to find node references (i.e. {"@id": "ref"}) that match a top-level object. The process can repeat until the number of remaining top-level objects remains the same. if it ends in a single object (as it typically will), the array containing that object can be replaced with just the object.
> 
> This certainly should be described in a sub-algorithm (call out). I'm happy to work on expressing such an algorithm during the course of my implementation.

I have also played with this algorithm yesterday. I tried to start fresh and came up with an approach that is probably slightly different than @gkellogg's or @JeniT's, though obviously bears similarities. Which bears the question: how much details do we want to put into the text for the algorithms, and how normative should that be? I must admit I do not like specs that are, essentially, commented codes...

My preferred approach is that a normative section describes what the algorithm should achieve calling out possible pitfalls (essentially, avoiding cycles, ie, building a forest), and have an informative section describing _a_ possible algorithm, and have an abundant number of test cases...

**Comment by iherman** [comment]
So, here is my take on the algorithm for nesting. It is more or less implemented although would require more testing on my side, but I think it works. Tastes differ but I personally find this formulation clearer. 
1. Objects in the current row should be merged when they share the same subjects (note that if two objects has a subject null, they should be considered to be different). The result is an array of objects, the _Object List_.
2. Make a list of URLs that appear as values of `aboutUrl`s in the row _only once_ (regardless of the name in the corresponding name/value pair). The result is the _URL List_.
3. Create a traditional "forest" data structure `F`: an array of _trees_ (i.e., directed acyclic graphs), with nodes containing a reference (e.g., the index in the _Object List_) for the current row objects. `F` is initialized to be empty. 
4. For each object `O` in the _Object List_ do as follows:
   1. Check if there is a node `N` in `F` that refers to `O`. If there isn't, create a new tree in the `F` with a new node `N` as a root, where `N` refers to `O`.
   2. Check all values in `O` that appear as `aboutUrl`s. If that URL appears in the _URL List_, then:
      1. Identify the object `V` whose identifier (i.e., the value of the name `@id`) is that URL;
      2. Check if the root of the tree, containing `N`, refers to `V`. If so, skip the remaining steps, i.e., go to 4.ii. and consider the next value in `O` (if any);
      3. Else if there is root node `M` in the tree that refers to `V`, make `M` a child of `N` (i.e., merge two trees);
      4. Else, create a new node `M` in the tree that refers to `V`, and make `M` a child of `N`.
5. Create a new forest `G` of JSON objects, mirroring the structure of `F`, as follows:
   1. each node `N<sub>G</sub>` is the object from _Object List_ referred to by the corresponding `N<sub>F</sub>` node;
   2. if `M<sub>F</sub>` is the child of `N<sub>F</sub>`, then, in `N<sub>G</sub>` the corresponding name/value pair (i.e., the `aboutUrl` that is at the origin of this relationship) is exchanged against the same name, but `M<sub>G</sub>`. 
6. Insert all the root objects in `G` in the final array `A`. 

Of course, implementations should probably optimize (e.g, if _URL list_ is empty, or the length of _Object List_ is one, then the whole process can be skipped), but that is besides the point. This is meant to give a direction and to dictate the detailed implementation...

**Comment by gkellogg** [comment]
To guard against the possibility of substituting an object with an @id which is the string value of another property, we may want to have the first pass of producing use a node object when there's a `valueUrl`, and then substitute only when the `@id` matches. We can then do the JSON-LD simplification step to the entire row result after nesting is done, rather than only doing it for common properties and notes.

There's also a question if the "describes" value should be an array with a single object, or just that single object. There's good arguments either way; JSON-LD typically compacts to using just an object value in this case, unless the term is defined using `@container: @set`.

Lastly, the last step of the existing CSV2JSON algorithm does not indicate how **A** is added to to **A<sub>R</sub>**; I presume that the array **A** is appended to **A<sub>R</sub>**, so that the final result will be an array of arrays; I don't think we have an existing example. Note that the tree-ops-virtual example would be appropriate for this.

**Comment by iherman** [comment]
> On 13 Mar 2015, at 24:50 , Gregg Kellogg notifications@github.com wrote:
> 
> To guard against the possibility of substituting an object with an @id which is the string value of another property, we may want to have the first pass of producing use a node object when there's a valueUrl, and then substitute only when the @id matches. We can then do the JSON-LD simplification step to the entire row result after nesting is done, rather than only doing it for common properties and notes.

@gkellogg I must admit I do not understand what you want to propose here. I have the feeling that the nesting algorithm, both [@JeniT's version](https://github.com/w3c/csvw/issues/299#issuecomment-77687103) and [mine](https://github.com/w3c/csvw/issues/299#issuecomment-78486015) do this.

But I may miss something... caveat: I am severely jet-lagged at the moment, got back home from North America early today...:-(

> There's also a question if the "describes" value should be an array with a single object, or just that single object. There's good arguments either way; JSON-LD typically compacts to using just an object value in this case, unless the term is defined using @container: @set.

I would go for the simple case: use an array in all cases, even if all the objects (columns) rely on the same subject.

(Admin: this is unrelated to the issues' original topic. If we want to get into this, let us use a different issue...)

> Lastly, the last step of the existing CSV2JSON algorithm does not indicate how A is added to to AR; I presume that the array A is appended to AR, so that the final result will be an array of arrays; I don't think we have an existing example. Note that the tree-ops-virtual example would be appropriate for this.

(Again, this should be a separate issue I believe...)
- In standard mode, (5.6.4) does say described 'A', so I am not sure what is missing in your view. Because that pair goes into 'R', which is part of 'AR', I think the definition is correct.
- In minimal mode, (1) defines 'A' on the top level

What am I missing?

Ivan

**Comment by gkellogg** [comment]
> @gkellogg I must admit I do not understand what you want to propose here. I have the feeling that the nesting algorithm, both [@JeniT's version](https://github.com/w3c/csvw/issues/299#issuecomment-77687103) and [mine](https://github.com/w3c/csvw/issues/299#issuecomment-78486015) do this.

Perhaps I'm not implementing to the algorithm right now, but if I have a column using valueUrl, I might get an output such as the following:

```
{ "dc:source": "http://example.org/foo"}
```

If I had a literal value that was the same, I'd get the same output. In RDF, I'd either get a literal or a URI. If you want to chain only when the value represents a URI and not a literal, you could start with the following:

```
{"dc:source": {"@id": "http://example.org/foo"}}
```

This would unambiguously identify this value as a URI, and not a literal. Admittedly, this is a narrow case, but one which JSON-LD goes to pains to deal with. Note that after JSON-LD simplification, the cumbersome `{"@id": ..}` would go away.

I'm just trying to point out this ambiguity; we can decide to go either way, but we should understand that it could possibly lead to unwarranted chaining.

**Comment by gkellogg** [comment]
> (Again, this should be a separate issue I believe...)
> - In standard mode, (5.6.4) does say described 'A', so I am not sure what is missing in your view. Because that pair goes into 'R', which is part of 'AR', I think the definition is correct.
> - In minimal mode, (1) defines 'A' on the top level

Didn't really want to create a separate issue unless it's warranted. It would be made more clear with an example. If I have AR defined as an empty array `[]`, and the result of processing an array is an array of objects `[{a}, {b}]`, is it appended to AR as [[{a},{b}]], or are the elements added to it? 5.6.4 describes standard mode, I'm concerned about minimal mode, which I thought was unclear.

**Comment by iherman** [comment]
> On 13 Mar 2015, at 17:47 , Gregg Kellogg notifications@github.com wrote:
> 
> @gkellogg I must admit I do not understand what you want to propose here. I have the feeling that the nesting algorithm, both @JeniT's version and mine do this.
> 
> Perhaps I'm not implementing to the algorithm right now, but if I have a column using valueUrl, I might get an output such as the following:
> 
> { "dc:source": "http://example.org/foo"}

Right.

> If I had a literal value that was the same, I'd get the same output. In RDF, I'd either get a literal or a URI. If you want to chain only when the value represents a URI and not a literal, you could start with the following:
> 
> {"dc:source": {"@id": "http://example.org/foo"}}
> 
> This would unambiguously identify this value as a URI, and not a literal. Admittedly, this is a narrow case, but one which JSON-LD goes to pains to deal with. Note that after JSON-LD simplification, the cumbersome {"@id": ..} would go away.

I think the point is: we look only at values that stem from a valueUrl processing. Ie, the algorithm, indeed, has to keep track of the 'origin' of the value, ie, an arbitrary equality of the values. I think this is important because we do not want to bring in a JSON-LD knowledge to this place (let that be done in the RDF case).

Ie, I am still not sure there is a problem

Ivan

> I'm just trying to point out this ambiguity; we can decide to go either way, but we should understand that it could possibly lead to unwarranted chaining.
> 
> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by iherman** [comment]
> On 13 Mar 2015, at 17:55 , Gregg Kellogg notifications@github.com wrote:
> 
> (Again, this should be a separate issue I believe...)
> 
>   • In standard mode, (5.6.4) does say described 'A', so I am not sure what is missing in your view. Because that pair goes into 'R', which is part of 'AR', I think the definition is correct.
>   • In minimal mode, (1) defines 'A' on the top level
> Didn't really want to create a separate issue unless it's warranted. It would be made more clear with an example. If I have AR defined as an empty array [], and the result of processing an array is an array of objects [{a}, {b}], is it appended to AR as [[{a},{b}]], or are the elements added to it? 5.6.4 describes standard mode, I'm concerned about minimal mode, which I thought was unclear.

Section 5.6.6 says, currently:

> For each subject i associated with the current row, recurse through the name-value pairs in the objects associated with that row to determine if subject i is referenced—excluding the object describing subject i itself (object Si).
> 
> If subject i is referenced once and only once, then object Si becomes the value of the referring name-value pair in place of the reference in order to create a nested object.
> 
> Else, insert object Si, potentially including nested objects, into array A.

The first two paragraph will have to be changed calling out to the separate section on the chaining algorithm, once we have decided which formulation we choose. But, for me, the last paragraph indicates that the result, in your example, should be [{a},{b}] in minimal mode; actually, there is no AR in minimal mode.

However... this made me realize that we may want to have a slightly more radical change of the formulation. While, in the RDF case, the minimal vs. standard mode is not terribly complicated because we just emit triples (or not), it is more complicated for the JSON case where the overall structure of the output is significantly different. I wonder whether it would not be cleaner to re-formulate the whole generation part by having two different sections, namely one for minimal mode and one for standard mode. (Possibly calling out to separated sections in case it makes sense, like what we intend to do for the nesting.) If even you get messed up, everyone will:-)

@6a6d74 , what do you think? I can try to do that at some point if we agree

Ivan

> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by iherman** [comment]
I have just picked up the idea in [my comment](https://github.com/w3c/csvw/issues/299#issuecomment-80531833) and made it so; see PR #362.

/Cc @6a6d74 @gkellogg

**Comment by iherman** [comment]
I have just pushed (in a separate branch!) a version of the PR #362 that includes the ["forest-tree" based description of the nesting algorithm](https://github.com/w3c/csvw/issues/299#issuecomment-78486015). The branch is `json-refactor-nesting`, can be best seen at: 

http://htmlpreview.github.io/?https://github.com/w3c/csvw/blob/csv2json-refactor-nesting/csv2json/index.html

If PR #362 is merged then I can make a pull request for this one, too (it is a branch of the version in #362)

/Cc @JeniT @6a6d74 @gkellogg

**Comment by gkellogg** [comment]
This looks right, but is not what I implement (not that that matters).

FYI, what I do more resembles the following:

When creating objects for each `subject`, record them in a map of `subject` to `object`.

When adding a the `value` of a valueUrl to an object `co` for predicate `p`, where `value` is not the subject of `co`, create a map from `value` to `co`, `p`, and a count, initially set to 1; if the map entry already exists, increase it's count by 1.

After processing all cells, for each `value`, `co`, `p`, and `count` where `count` is exactly 1, replace `value` in `co[p]` with the `object` from the subject map, where `subject` equals `value`, and remove the `subject/object` entry from that map.

The final result is all of the objects remaining in the `subject/object` map.

**Comment by iherman** [comment]
I am too tired and it is too late to think about this, but does it take care of the two extra complications: avoid cycles (I think it is o.k.) and ensure several layers of nesting (not sure at the moment)?

The reason I liked my approach is because what you really build, through nesting, _is_ a forest. Ie, it looked like a good idea to make that clear.

I still believe that the algorithm itself should be made non normative, ie, people should be allowed to choose an approach they prefer.

Ivan

---

Ivan Herman
Tel:+31 641044153
http://www.ivan-herman.net

(Written on mobile, sorry for brevity and misspellings...)

**Comment by gkellogg** [comment]
I think that the algorithm can be normative, as long as we make clear that it is the result of the algorithm that is important, not the actual steps followed. A agree that JSON-LD over-specified this, but I was out-voted. That said, I think your algorithm results in the same output. The way mine is stated is probably too programming-language specific.

**Comment by iherman** [comment]
> On 16 Mar 2015, at 05:15 , Gregg Kellogg notifications@github.com wrote:
> 
> I think that the algorithm can be normative, as long as we make clear that it is the result of the algorithm that is important, not the actual steps followed.

O.k. I have a note in the PR, maybe we can make it even more explicit.

> A agree that JSON-LD over-specified this, but I was out-voted. That said, I think your algorithm results in the same output.

It should:-)

Ivan

> The way mine is stated is probably too programming-language specific.
> 
> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by 6a6d74** [comment]
fixed in PR #369
